# Data Analysing for used car listing in Ebay using Numpy and Pandas

In this project,we will do data analysis on used car listing in Ebay, a section on ebay website. We are working with the sample 50,000 datasets.

The attributes in dataset is given below:

dateCrawled - When this ad was first crawled.<br>
name - Name of the car.<br>
seller - Whether the seller is private or a dealer.<br>
offerType - The type of listing<br>
price - The price on the ad to sell the car.<br>
abtest - Whether the listing is included in an A/B test.<br>
vehicleType - The vehicle Type.<br>
yearOfRegistration - The year in which which year the car was first registered.<br>
gearbox - The transmission type.<br>
powerPS - The power of the car in PS.<br>
model - The car model name.<br>
kilometer - How many kilometers the car has driven.<br>
monthOfRegistration - The month in which which year the car was first registered.<br>
fuelType - What type of fuel the car uses.<br>
brand - The brand of the car.<br>
notRepairedDamage - If the car has a damage which is not yet repaired.<br>
dateCreated - The date on which the eBay listing was created.<br>
nrOfPictures - The number of pictures in the ad.<br>
postalCode - The postal code for the location of the vehicle.<br>
lastSeenOnline - When the crawler saw this ad last online.<br>

The aim of this project is to data cleaning and data analysis using Numpy and Pandas.

In [1]:
import pandas as pd
import numpy as np

In [2]:
auto = pd.read_csv('C:\\Users\\archit.agarwal2\\Desktop\\Data\\auto.csv', encoding='Latin-1')
auto.info()
auto.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

dateCrawled                                               name  \
0  2016-03-24 11:52:17                                         Golf_3_1.6   
1  2016-03-24 10:58:45                               A5_Sportback_2.7_Tdi   
2  2016-03-14 12:52:21                     Jeep_Grand_Cherokee_"Overland"   
3  2016-03-17 16:54:04                                 GOLF_4_1_4__3TÜRER   
4  2016-03-31 17:25:20                     Skoda_Fabia_1.4_TDI_PD_Classic   
5  2016-04-04 17:36:23  BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...   
6  2016-04-01 20:48:51                        Peugeot_206_CC_110_Platinum   
7  2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
8  2016-04-04 23:42:13               Ford_C___Max_Titanium_1_0_L_EcoBoost   
9  2016-03-17 10:53:50  VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...   

   seller offerType  price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot    480     test         NaN                1993   
1  privat   Angebot  18300     test       coupe                2011   
2  privat   Angebot   9800     test         suv                2004   
3  privat   Angebot   1500     test  kleinwagen                2001   
4  privat   Angebot   3600     test  kleinwagen                2008   
5  privat   Angebot    650     test   limousine                1995   
6  privat   Angebot   2200     test      cabrio                2004   
7  privat   Angebot      0     test   limousine                1980   
8  privat   Angebot  14500  control         bus                2014   
9  privat   Angebot    999     test  kleinwagen                1998   

     gearbox  powerPS    model  kilometer  monthOfRegistration fuelType  \
0    manuell        0     golf     150000                    0   benzin   
1    manuell      190      NaN     125000                    5   diesel   
2  automatik      163    grand     125000                    8   diesel   
3    manuell       75     golf     150000                    6   benzin   
4    manuell       69    fabia      90000                    7   diesel   
5    manuell      102      3er     150000                   10   benzin   
6    manuell      109  2_reihe     150000                    8   benzin   
7    manuell       50   andere      40000                    7   benzin   
8    manuell      125    c_max      30000                    8   benzin   
9    manuell      101     golf     150000                    0      NaN   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0  volkswagen               NaN  2016-03-24 00:00:00             0   
1        audi                ja  2016-03-24 00:00:00             0   
2        jeep               NaN  2016-03-14 00:00:00             0   
3  volkswagen              nein  2016-03-17 00:00:00             0   
4       skoda              nein  2016-03-31 00:00:00             0   
5         bmw                ja  2016-04-04 00:00:00             0   
6     peugeot              nein  2016-04-01 00:00:00             0   
7  volkswagen              nein  2016-03-21 00:00:00             0   
8        ford               NaN  2016-04-04 00:00:00             0   
9  volkswagen               NaN  2016-03-17 00:00:00             0   

   postalCode             lastSeen  
0       70435  2016-04-07 03:16:57  
1       66954  2016-04-07 01:46:50  
2       90480  2016-04-05 12:47:46  
3       91074  2016-03-17 17:40:17  
4       60437  2016-04-06 10:17:21  
5       33775  2016-04-06 19:17:07  
6       67112  2016-04-05 18:18:39  
7       19348  2016-03-25 16:47:58  
8       94505  2016-04-04 23:42:13  
9       27472  2016-03-31 17:17:06

Our dataset contain 20 columns and most of the dataset is string. There are many null values too. We will start by cleaning the columns.

In [3]:
auto.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

First, we will make every columns name in Snakecase

In [4]:
auto.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
auto.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00           0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00           0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00           0        60437  2016-04-06 10:17:21

## Initail Data Exploration and cleaning|

In [5]:
auto.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Our Initial Observation

There are some columns where most of the values are same:
1. seller
2. offer_type

We need to explore the num_photos columns as it look odd.

In [6]:
auto["num_photos"].value_counts()

0    371528
Name: num_photos, dtype: int64

As it looks all the values in num_photos is 0, so we will remove this column plus the other two namely seller and offer_type

In [7]:
auto = auto.drop(["num_photos", "seller", "offer_type"], axis=1)

There are two columns which, price and odometer, are numeric values but stored as text and it have also extra characters.

In [8]:
auto["price"] = (auto["price"].replace("$","").replace(",","").astype(int))
auto["price"].head()

0      480
1    18300
2     9800
3     1500
4     3600
Name: price, dtype: int32

In [9]:
auto["odometer"] = (auto["odometer"].replace("km","").replace(",","").astype(int)                            )
auto.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
auto["odometer_km"].head()

0    150000
1    125000
2    125000
3    150000
4     90000
Name: odometer_km, dtype: int32

### Exploring Odometer and price

In [10]:
auto["odometer_km"].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

We can see that values in Odometer field is rounded off which might seems that sellers have to choose preset option from the list

In [11]:
print(auto["price"].unique().shape)
print(auto["price"].describe())
auto["price"].value_counts().head(20)

(5597,)
count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64

Again , It seems to be rounded off the price. Here, we can see that 10778 cars are listed with $0 which is invalid which we will remove from the list. The maximum price of the car is 2 millions dolars which seems to be non obvious, we will remove that value too.

In [12]:
Value_count = auto["price"].value_counts()
Sort_value = Value_count.sort_index(ascending=False)
first_20 = Sort_value.head(40)
print(first_20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
1111111        2
1010010        1
1000000        5
999999        13
999990         1
911911         1
849000         1
820000         1
780000         1
745000         2
725000         1
700000         1
650000         1
619000         1
600000         2
599000         1
585000         1
579000         1
517895         1
500000         2
Name: price, dtype: int64


In [13]:
Value_count = auto["price"].value_counts()
Sort_value = Value_count.sort_index(ascending=True)
first_20 = Sort_value.head(20)
print(first_20)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64


We can see that there are many cars above around 1 million dollor and many cars are listed below 1000 dolors. We will remove all these values above 50,0000 doolar and below 1000 dollar

In [14]:
auto = auto[auto["price"].between(1000,500000)]
auto["price"].describe()

count    288093.000000
mean       7260.710545
std        9955.801724
min        1000.000000
25%        2200.000000
50%        4350.000000
75%        8950.000000
max      500000.000000
Name: price, dtype: float64

## Exploring the date columns 

There are a number of columns with date information:

1. date_crawled
2. registration_month
3. registration_year
4. ad_created
5. last_seen

We will explore each of the columns:

In [15]:
auto[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21
6  2016-04-01 20:48:51  2016-04-01 00:00:00  2016-04-05 18:18:39

In [16]:
(auto["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025835
2016-03-06    0.014683
2016-03-07    0.035214
2016-03-08    0.032969
2016-03-09    0.033468
2016-03-10    0.032542
2016-03-11    0.032757
2016-03-12    0.036613
2016-03-13    0.016082
2016-03-14    0.036370
2016-03-15    0.032507
2016-03-16    0.029699
2016-03-17    0.030990
2016-03-18    0.012944
2016-03-19    0.035489
2016-03-20    0.036644
2016-03-21    0.035124
2016-03-22    0.032042
2016-03-23    0.031799
2016-03-24    0.029851
2016-03-25    0.032944
2016-03-26    0.032750
2016-03-27    0.030855
2016-03-28    0.035294
2016-03-29    0.033861
2016-03-30    0.033166
2016-03-31    0.031507
2016-04-01    0.034701
2016-04-02    0.035981
2016-04-03    0.039751
2016-04-04    0.037970
2016-04-05    0.012763
2016-04-06    0.003211
2016-04-07    0.001624
Name: date_crawled, dtype: float64

Look like site crawled daily in March and April 2016

In [17]:
(auto["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001107
2016-03-06    0.003652
2016-03-07    0.004655
2016-03-08    0.007310
2016-03-09    0.008803
2016-03-10    0.010451
2016-03-11    0.012152
2016-03-12    0.021746
2016-03-13    0.007807
2016-03-14    0.011659
2016-03-15    0.015030
2016-03-16    0.015443
2016-03-17    0.027092
2016-03-18    0.006574
2016-03-19    0.015433
2016-03-20    0.018758
2016-03-21    0.018935
2016-03-22    0.019542
2016-03-23    0.017085
2016-03-24    0.018272
2016-03-25    0.017970
2016-03-26    0.015106
2016-03-27    0.015797
2016-03-28    0.021122
2016-03-29    0.022284
2016-03-30    0.022406
2016-03-31    0.022993
2016-04-01    0.023288
2016-04-02    0.024301
2016-04-03    0.024575
2016-04-04    0.024336
2016-04-05    0.133460
2016-04-06    0.232165
2016-04-07    0.138691
Name: last_seen, dtype: float64

The last saw any listing, allows us to determine on what day a listing was removed, presumably because the car was sold.

In [18]:
print(auto["ad_created"].str[:10].unique().shape)
(auto["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(112,)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
2015-08-10    0.000003
2015-09-09    0.000003
2015-10-14    0.000003
2015-11-02    0.000003
2015-11-08    0.000003
2015-11-10    0.000003
2015-11-12    0.000003
2015-11-13    0.000003
2015-11-17    0.000003
2015-11-23    0.000007
2015-11-24    0.000007
2015-12-06    0.000003
2015-12-17    0.000003
2015-12-27    0.000003
2015-12-28    0.000003
2015-12-30    0.000010
2016-01-02    0.000003
2016-01-03    0.000003
2016-01-06    0.000003
2016-01-07    0.000010
2016-01-08    0.000003
2016-01-10    0.000014
2016-01-13    0.000010
2016-01-14    0.000003
2016-01-15    0.000003
                ...   
2016-03-09    0.033562
2016-03-10    0.032378
2016-03-11    0.032812
2016-03-12    0.036412
2016-03-13    0.017460
2016-03-14    0.035152
2016-03-15    0.032420
2016-03-16    0.029952
2016-03-17    0.030594
2016-03-18    0.013877
2016-03-19    0.034472
2016-03-20    0.036717
2016-03-21 

There are large variety of ads created dates

In [19]:
auto["registration_year"].describe()

count    288093.000000
mean       2005.125491
std          63.806135
min        1000.000000
25%        2001.000000
50%        2005.000000
75%        2009.000000
max        9999.000000
Name: registration_year, dtype: float64

From the registered age, we can figure out the age of the car. We can see that min: registered year is 1000(long back when even cars not invented) and max is 9999 (into the future). So we need to clean this invalid data.

## Dealing with invalid registration data

Let's determine what percentage of our data has invalid values in this column:

In [20]:
(~auto["registration_year"].between(1900,2016)).sum() / auto.shape[0]

0.03743929911521627

Comes out to be less than 4%, we will remove this data

In [21]:
auto = auto[auto["registration_year"].between(1900,2016)]
auto["registration_year"].value_counts(normalize=True).head(10)

2005    0.074380
2006    0.071369
2004    0.067820
2003    0.065123
2007    0.062638
2002    0.059151
2008    0.057572
2001    0.055873
2009    0.055693
2000    0.053136
Name: registration_year, dtype: float64

It comes out to be that most of the vehicles registered 20 years ago

## Exploring Price by Brand

In [22]:
auto["brand"].value_counts(normalize=True)

volkswagen        0.212036
bmw               0.124948
mercedes_benz     0.111346
audi              0.100556
opel              0.087524
ford              0.057828
renault           0.037114
peugeot           0.028827
fiat              0.020948
skoda             0.018380
seat              0.017807
smart             0.016484
toyota            0.014799
mazda             0.014533
citroen           0.013653
nissan            0.012052
mini              0.011626
hyundai           0.010797
sonstige_autos    0.010032
volvo             0.008748
porsche           0.007605
kia               0.007504
honda             0.007162
mitsubishi        0.006632
suzuki            0.006177
alfa_romeo        0.006109
chevrolet         0.005885
chrysler          0.003700
dacia             0.003054
land_rover        0.002690
jeep              0.002676
jaguar            0.002095
subaru            0.001745
saab              0.001410
daihatsu          0.001388
trabant           0.000959
lancia            0.000916
d

It seems that volkswagen is most popular brand with approximately double the cars for sale of the next two brands combined..German manufacturers represent four out of the top five brands in most listed cars.

There are lots of brands that have total less than 5% listings ,so we will limit our analysis to brands representing more than 5% of total listings.

In [23]:
brand_counts = auto["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford'], dtype='object')


In [24]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = auto[auto["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 6597,
 'bmw': 9240,
 'mercedes_benz': 9251,
 'audi': 10045,
 'opel': 4224,
 'ford': 5241}

### We can see analysis from the data that
* audi, mercedez_benz and bmw are being the most expensive car 
* opel and ford is lowest
* Volkswagen is between two extremeties this may explain its popularity

## Exploring Milage


In [25]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = auto[auto["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

In [26]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132174
mercedes_benz        129767
audi                 128036
volkswagen           125502
opel                 123021
ford                 119242

In [27]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132174        9240
mercedes_benz        129767        9251
audi                 128036       10045
volkswagen           125502        6597
opel                 123021        4224
ford                 119242        5241

The range of the car does not vary as much like the price of the car. But it seems to be that expensive cars has likely to be have more mileage than less expensive cars.